src/node_file.cc

```diff
@@ -2727,6 +2730,238 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+  return true;
+#endif
+}
+
+FilePathIsFileReturnType FilePathIsFile(Environment* env,
+                                        const std::string& file_path) {
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env,
+      permission::PermissionScope::kFileSystemRead,
+      file_path,
+      FilePathIsFileReturnType::kThrowInsufficientPermissions);
+
+  uv_fs_t req;
+
+  int rc = uv_fs_stat(env->event_loop(), &req, file_path.c_str(), nullptr);
+
+  if (rc == 0) {
+    const uv_stat_t* const s = static_cast<const uv_stat_t*>(req.ptr);
+    rc = !!(s->st_mode & S_IFDIR);
+  }
+
+  uv_fs_req_cleanup(&req);
+
+  // rc is 0 if the path refers to a file
+  if (rc == 0) return FilePathIsFileReturnType::kIsFile;
+
+  return FilePathIsFileReturnType::kIsNotFile;
+}
+
+static void LegacyMainResolve(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 1);
+  // packageJSONUrl
+  // /home/user/project/package.json
+  CHECK(args[0]->IsString());
+
+  Environment* env = Environment::GetCurrent(args);
+
+  Utf8Value utf8_base_path = Utf8Value(env->isolate(), args[0].As<String>());
+  auto base = ada::parse<ada::url_aggregator>(utf8_base_path.ToStringView());
+
+  if (!base) {
+    return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+  }
+
+  ada::result<ada::url_aggregator> file_path_url;
+  std::string initial_file_path;
+  std::string file_path;
+
+  // packageConfig.main
+  // ./index.js
+  // ./lib/index.js
+  if (args[1]->IsString()) {
+    std::string package_config_main =
+        Utf8Value(env->isolate(), args[1].As<String>()).ToString();
+
+    file_path_url = ada::parse<ada::url_aggregator>(
+        std::string("./") + package_config_main, &base.value());
+
+    if (!file_path_url) {
+      return args.GetReturnValue().Set(
+          LegacyMainResolveReturnType::kInvalidUrl);
+    }
+
+    if (!FileURLToPathImpl(env, file_path_url.value(), initial_file_path))
+      return;
+
+    FromNamespacedPath(&initial_file_path);
+
+    for (int i = 0; i < legacy_main_extensions_with_main_end; i++) {
+      file_path = initial_file_path + legacy_main_extensions[i];
+
+      switch (FilePathIsFile(env, file_path)) {
+        case FilePathIsFileReturnType::kIsFile:
+          return args.GetReturnValue().Set(i);
+        case FilePathIsFileReturnType::kIsNotFile:
+          continue;
+        case FilePathIsFileReturnType::kThrowInsufficientPermissions:
+          // the default behavior when do not have permission is to return
+          // and exit the execution of the method as soon as possible
+          // the internal function will throw the exception
+          return;
+        default:
+          UNREACHABLE();
+      }
+    }
+  }
+
+  file_path_url = ada::parse<ada::url_aggregator>("./index", &base.value());
+
+  if (!file_path_url) {
+    return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+  }
+
+  if (!FileURLToPathImpl(env, file_path_url.value(), initial_file_path)) return;
```

Comment:
```suggestion
  if (!FileURLToPath(env, file_path_url.value(), initial_file_path)) return;
```

Reply:
The idea of using `Impl` in the end was because of that method could be used to implement the real `fileURLToPath` on the JS side.

But after the benchmarks, probably doesn't worth having this method exported.